*** /usr/local/lib/perl5/5.10.1/SelfLoader.pm	Fri Sep 17 15:44:38 2010
--- /users/mike/SelfLoader.pm	Mon Jan 10 16:50:13 2011
***************
*** 1,7 ****
  package SelfLoader;
  use 5.008;
  use strict;
! our $VERSION = "1.17";
  
  # The following bit of eval-magic is necessary to make this work on
  # perls < 5.009005.
--- 1,8 ----
  package SelfLoader;
  use 5.008;
  use strict;
! use IO::Handle;
! our $VERSION = "1.18";
  
  # The following bit of eval-magic is necessary to make this work on
  # perls < 5.009005.
***************
*** 102,107 ****
--- 103,109 ----
        close $fh or die "close: $!";                 # autocloses, but be paranoid
        open $fh, '<&', $nfh or croak "reopen2: $!";  # dup() the fd "back"
        close $nfh or die "close after reopen: $!";   # autocloses, but be paranoid
+       $fh->untaint;
      }
      $Cache{"${currpack}::<DATA"} = 1;   # indicate package is cached
  
