diff -ur tcsh-6.14.00.orig/config.sub tcsh-6.14.00/config.sub
--- tcsh-6.14.00.orig/config.sub	2002-07-12 07:14:20.000000000 -0600
+++ tcsh-6.14.00/config.sub	2010-07-21 12:51:55.000000000 -0600
@@ -1216,6 +1216,9 @@
 		;;
 	-none)
 		;;
+	-uclibc)
+		os=-linux-gnu
+		;;
 	*)
 		# Get rid of the `-' at the beginning of $os.
 		os=`echo $os | sed 's/[^-]*-//'`
diff -ur tcsh-6.14.00.orig/configure tcsh-6.14.00/configure
--- tcsh-6.14.00.orig/configure	2005-03-23 06:20:30.000000000 -0700
+++ tcsh-6.14.00/configure	2010-07-21 13:33:28.000000000 -0600
@@ -7110,72 +7110,75 @@
 
 fi
 
-echo "$as_me:$LINENO: checking whether setpgrp takes no argument" >&5
-echo $ECHO_N "checking whether setpgrp takes no argument... $ECHO_C" >&6
-if test "${ac_cv_func_setpgrp_void+set}" = set; then
-  echo $ECHO_N "(cached) $ECHO_C" >&6
-else
-  if test "$cross_compiling" = yes; then
-  { { echo "$as_me:$LINENO: error: cannot check setpgrp when cross compiling" >&5
-echo "$as_me: error: cannot check setpgrp when cross compiling" >&2;}
-   { (exit 1); exit 1; }; }
-else
-  cat >conftest.$ac_ext <<_ACEOF
-/* confdefs.h.  */
-_ACEOF
-cat confdefs.h >>conftest.$ac_ext
-cat >>conftest.$ac_ext <<_ACEOF
-/* end confdefs.h.  */
-#if HAVE_UNISTD_H
-# include <unistd.h>
-#endif
-
-int
-main ()
-{
-/* If this system has a BSD-style setpgrp which takes arguments,
-  setpgrp(1, 1) will fail with ESRCH and return -1, in that case
-  exit successfully. */
-  exit (setpgrp (1,1) == -1 ? 0 : 1);
-  ;
-  return 0;
-}
-_ACEOF
-rm -f conftest$ac_exeext
-if { (eval echo "$as_me:$LINENO: \"$ac_link\"") >&5
-  (eval $ac_link) 2>&5
-  ac_status=$?
-  echo "$as_me:$LINENO: \$? = $ac_status" >&5
-  (exit $ac_status); } && { ac_try='./conftest$ac_exeext'
-  { (eval echo "$as_me:$LINENO: \"$ac_try\"") >&5
-  (eval $ac_try) 2>&5
-  ac_status=$?
-  echo "$as_me:$LINENO: \$? = $ac_status" >&5
-  (exit $ac_status); }; }; then
-  ac_cv_func_setpgrp_void=no
-else
-  echo "$as_me: program exited with status $ac_status" >&5
-echo "$as_me: failed program was:" >&5
-sed 's/^/| /' conftest.$ac_ext >&5
+#echo "$as_me:$LINENO: checking whether setpgrp takes no argument" >&5
+#echo $ECHO_N "checking whether setpgrp takes no argument... $ECHO_C" >&6
+#if test "${ac_cv_func_setpgrp_void+set}" = set; then
+#  echo $ECHO_N "(cached) $ECHO_C" >&6
+#else
+#  if test "$cross_compiling" = yes; then
+#  { { echo "$as_me:$LINENO: error: cannot check setpgrp when cross compiling" >&5
+#echo "$as_me: error: cannot check setpgrp when cross compiling" >&2;}
+#   { (exit 1); exit 1; }; }
+#else
+#  cat >conftest.$ac_ext <<_ACEOF
+#/* confdefs.h.  */
+#_ACEOF
+#cat confdefs.h >>conftest.$ac_ext
+#cat >>conftest.$ac_ext <<_ACEOF
+#/* end confdefs.h.  */
+##if HAVE_UNISTD_H
+## include <unistd.h>
+##endif
+
+#int
+#main ()
+#{
+#/* If this system has a BSD-style setpgrp which takes arguments,
+#  setpgrp(1, 1) will fail with ESRCH and return -1, in that case
+#  exit successfully. */
+#  exit (setpgrp (1,1) == -1 ? 0 : 1);
+#  ;
+#  return 0;
+#}
+#_ACEOF
+#rm -f conftest$ac_exeext
+#if { (eval echo "$as_me:$LINENO: \"$ac_link\"") >&5
+#  (eval $ac_link) 2>&5
+#  ac_status=$?
+#  echo "$as_me:$LINENO: \$? = $ac_status" >&5
+#  (exit $ac_status); } && { ac_try='./conftest$ac_exeext'
+#  { (eval echo "$as_me:$LINENO: \"$ac_try\"") >&5
+#  (eval $ac_try) 2>&5
+#  ac_status=$?
+#  echo "$as_me:$LINENO: \$? = $ac_status" >&5
+#  (exit $ac_status); }; }; then
+#  ac_cv_func_setpgrp_void=no
+#else
+#  echo "$as_me: program exited with status $ac_status" >&5
+#echo "$as_me: failed program was:" >&5
+#sed 's/^/| /' conftest.$ac_ext >&5
+
+#( exit $ac_status )
+#ac_cv_func_setpgrp_void=yes
+#fi
+#rm -f core *.core gmon.out bb.out conftest$ac_exeext conftest.$ac_objext conftest.$ac_ext
+#fi
+#fi
+#echo "$as_me:$LINENO: result: $ac_cv_func_setpgrp_void" >&5
+#echo "${ECHO_T}$ac_cv_func_setpgrp_void" >&6
+#if test $ac_cv_func_setpgrp_void = yes; then
+
+#cat >>confdefs.h <<\_ACEOF
+##define SETPGRP_VOID 1
+#_ACEOF
 
-( exit $ac_status )
-ac_cv_func_setpgrp_void=yes
-fi
-rm -f core *.core gmon.out bb.out conftest$ac_exeext conftest.$ac_objext conftest.$ac_ext
-fi
-fi
-echo "$as_me:$LINENO: result: $ac_cv_func_setpgrp_void" >&5
-echo "${ECHO_T}$ac_cv_func_setpgrp_void" >&6
-if test $ac_cv_func_setpgrp_void = yes; then
+#fi
 
+# XXX hack
 cat >>confdefs.h <<\_ACEOF
 #define SETPGRP_VOID 1
 _ACEOF
 
-fi
-
-
-
