--- kexec-tools/util/Makefile.orig	2008-08-19 13:43:45.000000000 -0600
+++ kexec-tools/util/Makefile	2008-08-19 13:43:58.000000000 -0600
@@ -4,8 +4,8 @@
 	@$(MKDIR) -p $(@D)
 	$(LINK.o) -o $@ $^
 
-$(BIN_TO_HEX): CC=$(BUILD_CC)
-$(BIN_TO_HEX): CFLAGS=$(BUILD_CFLAGS)
+$(BIN_TO_HEX): CC=$(CC_FOR_BUILD)
+$(BIN_TO_HEX): CFLAGS=$(CFLAGS_FOR_BUILD)
 
 dist += util/Makefile util/bin-to-hex.c
 clean += util/bin-to-hex.o $(BIN_TO_HEX)
