Index: Makefile
===================================================================
RCS file: /share/freebsd/CVS/ports/misc/Makefile,v
retrieving revision 1.918
diff -u -r1.918 Makefile
--- Makefile    29 Jun 2009 12:04:10 -0000      1.918
+++ Makefile    7 Jan 2010 23:14:23 -0000
@@ -561,5 +561,23 @@
     SUBDIR += ytree
     SUBDIR += zaptel
     SUBDIR += zoneinfo
+# Emulab ports
+    SUBDIR += elvind
+    SUBDIR += emulab-boss
+    SUBDIR += emulab-fs
+    SUBDIR += emulab-ops
+    SUBDIR += emulab-clrhouse
+    SUBDIR += emulab-protogeni
+    SUBDIR += libelvin
+    SUBDIR += emulab-tftp-hpa
+    SUBDIR += p5-GD-old
+    SUBDIR += p5-DBD-mysql50-old
+    SUBDIR += ulsshxmlrpcpp
+    SUBDIR += emulab-iperf
+    SUBDIR += emulab-php4
+    SUBDIR += emulab-mailman
+    SUBDIR += emulab-pubsub
+    SUBDIR += emulab-pubsub-elvincompat
+    SUBDIR += emulab-py-m2crypto
 
 .include <bsd.port.subdir.mk>
