Only in tg2.0/src/tg: #prot_dgram.c#
diff -ru dist-tg2.0/src/tg/log.c tg2.0/src/tg/log.c
--- dist-tg2.0/src/tg/log.c	Thu Jan 24 16:30:09 2002
+++ tg2.0/src/tg/log.c	Wed Mar 13 12:43:19 2002
@@ -268,6 +268,11 @@
 	char		*cp = buf;
 	char		ctl;
 
+#ifdef USEEVENTS
+	if (log_fp == NULL)
+		return;
+#endif
+
 	/* Encode record type.						*/
 
 	*cp++ = LOGTYPE_ACCEPT;
@@ -326,6 +331,11 @@
 	char		*cp = buf;
 	char		ctl;
 
+#ifdef USEEVENTS
+	if (log_fp == NULL)
+		return;
+#endif
+
 	/* Encode record type.						*/
 
 	*cp++ = LOGTYPE_ERROR;
@@ -391,6 +401,11 @@
 	char		*cp = buf;
 	char		ctl;
 
+#ifdef USEEVENTS
+	if (log_fp == NULL)
+		return;
+#endif
+
 	/* Encode record type.						*/
 
 	*cp++ = LOGTYPE_RX;
@@ -459,6 +474,11 @@
 	char		*cp = buf;
 	char		ctl;
 
+#ifdef USEEVENTS
+	if (log_fp == NULL)
+		return;
+#endif
+
 	/* Encode record type.						*/
 
 	*cp++ = LOGTYPE_SETUP;
@@ -510,6 +530,11 @@
 	char		*cp = buf;
 	char		ctl;
 
+#ifdef USEEVENTS
+	if (log_fp == NULL)
+		return;
+#endif
+
 	/* Encode record type.						*/
 
 	*cp++ = LOGTYPE_TEARDOWN;
@@ -567,6 +592,11 @@
 	char		buf[100];
 	char		*cp = buf;
 	char		ctl;
+
+#ifdef USEEVENTS
+	if (log_fp == NULL)
+		return;
+#endif
 
 	/* Encode record type.						*/
 
diff -ru dist-tg2.0/src/tg/prot_dgram.c tg2.0/src/tg/prot_dgram.c
--- dist-tg2.0/src/tg/prot_dgram.c	Thu Jan 24 16:30:10 2002
+++ tg2.0/src/tg/prot_dgram.c	Wed Mar 20 10:20:15 2002
@@ -198,6 +198,21 @@
 	for (;;)
 		{
 
+#ifdef USEEVENTS
+		extern int gotevent;
+		tgevent_poll();
+		if (gotevent)
+			{
+			rcving = 0;
+			errno = ETIME;
+			return (0);
+			}
+		if (tv.tv_sec > 0 || tv.tv_usec > 100000) {
+			tv.tv_sec = 0;
+			tv.tv_usec = 100000;
+		}
+#endif
+
 		/* Set up for select:  get fd bitmaps.			*/
 
 		rfds = fds;
@@ -427,6 +442,7 @@
 
 			if (fromlen == 0)
 				{
+				(*(prtab->buffer_free))(buf);
 				errno = EBADF;
 				return (-1);
 				}
@@ -489,6 +505,7 @@
 
 	if (!dgram_get_packets(fd, endtout))
 		{
+		(*(prtab->buffer_free))(buf);
 		return (-1);
 		}
 	else
diff -ru dist-tg2.0/src/tg/prot_stream.c tg2.0/src/tg/prot_stream.c
--- dist-tg2.0/src/tg/prot_stream.c	Thu Jan 24 16:30:10 2002
+++ tg2.0/src/tg/prot_stream.c	Wed Mar 20 10:21:08 2002
@@ -147,6 +147,19 @@
 	for (;;)
 		{
 
+#ifdef USEEVENTS
+		extern int gotevent;
+		tgevent_poll();
+		if (gotevent)
+			{
+			rcving = 0;
+			errno = ETIME;
+			return (0);
+			}
+		tv.tv_sec = 0;
+		tv.tv_usec = 10000;
+#endif
+
 		/* Set up for select:  get fd bitmaps.			*/
 
 		rfds = fds;
@@ -440,6 +453,7 @@
 	    /* new connections while we are waiting to write.		*/
 
             if (!stream_get_packets(fd, endtout)) {
+		buffer_generic_free(buf);
 		return (-1);
             } else {
 
diff -ru dist-tg2.0/src/tg/prot_tcp.c tg2.0/src/tg/prot_tcp.c
--- dist-tg2.0/src/tg/prot_tcp.c	Thu Jan 24 16:30:10 2002
+++ tg2.0/src/tg/prot_tcp.c	Wed Mar 13 09:16:49 2002
@@ -184,7 +184,7 @@
 			return (-1);
 		}	
 
-		if ((connect(sfd, &(prot->dst), sizeof(prot->dst)) < 0) &
+		if ((connect(sfd, &(prot->dst), sizeof(prot->dst)) < 0) &&
 		    (errno != EINPROGRESS))
 			{
 			(void)close(sfd);
diff -ru dist-tg2.0/src/tg/prot_test.c tg2.0/src/tg/prot_test.c
--- dist-tg2.0/src/tg/prot_test.c	Thu Jan 24 16:30:10 2002
+++ tg2.0/src/tg/prot_test.c	Wed Mar 20 09:11:30 2002
@@ -447,6 +447,7 @@
 
 	if (!test_get_packets(prtab, fd, endtout))
 		{
+		buffer_generic_free(buf);
 		return (-1);
 		}
 	else
diff -ru dist-tg2.0/src/tg/prot_udp.c tg2.0/src/tg/prot_udp.c
--- dist-tg2.0/src/tg/prot_udp.c	Thu Jan 24 16:30:10 2002
+++ tg2.0/src/tg/prot_udp.c	Wed Mar 13 09:16:49 2002
@@ -205,7 +205,7 @@
 				return (-1);
 				}
 			}
-		if ((connect(sfd, &(prot->dst), sizeof(prot->dst)) < 0) &
+		if ((connect(sfd, &(prot->dst), sizeof(prot->dst)) < 0) &&
 		    (errno != EINPROGRESS))
 			{
 			(void)close(sfd);
@@ -232,7 +232,7 @@
 
 
 
-		    /* don't think this is correect */
+		    /* don't think this is correct */
 		addr = (struct sockaddr_in *)&(prot->src);
 		if ( IN_MULTICAST(ntohl(addr->sin_addr.s_addr)))		
 	        {
diff -ru dist-tg2.0/src/tg/tg.y tg2.0/src/tg/tg.y
--- dist-tg2.0/src/tg/tg.y	Thu Jan 24 16:30:10 2002
+++ tg2.0/src/tg/tg.y	Wed Mar 13 09:16:49 2002
@@ -133,7 +133,10 @@
 char                    *version = "2.0";	/* TG program version.	*/
 char			*ofile = NULL;
 char			*ifile = NULL;
- int			FlushOutput = 0;      /* whether to flush after each*/
+int			FlushOutput = 0;      /* whether to flush after each*/
+#ifdef USEEVENTS
+int			gotevent;
+#endif
 
 
 /* Variables local to this file.					*/
@@ -856,6 +859,10 @@
 	void		sigint();
 	FILE		*fp;
 
+#ifdef USEEVENTS
+	tgevent_init(argc, argv);
+	tgevent_loop();
+#else
 	/* Set debugging if it is desired.				*/
 
 #	if YYDEBUG
@@ -917,6 +924,7 @@
 	/* Generate traffic.						*/
 
 	do_actions();
+#endif
 	return (0);
 	}
 
@@ -1017,6 +1025,10 @@
 	/* Each pass through the following loop processes one tg_action	*/
 	/* element from the list.					*/
 
+#ifdef USEEVENTS
+ restart:
+	gotevent = 0;
+#endif
 	for (cur_tg = tg_first; cur_tg != NULL; cur_tg = cur_tg->next)
 		{
 
@@ -1116,7 +1128,7 @@
 
 			(*(prot.prot->sleep_till))(&(cur_tg->stop_before));
 			}
-	else if ((cur_tg->tg_flags & TG_LOG) != 0)
+		else if ((cur_tg->tg_flags & TG_LOG) != 0)
 			{
 			/* (Re-)Start logging to a new file */
 			time_t now;
@@ -1144,6 +1156,10 @@
 
 			generate(tx_asn, cur_tg, lasttime);
 			}
+#ifdef USEEVENTS
+		if (gotevent)
+			goto restart;
+#endif
 		}
 
 	/* Finished, tear down connection.				*/
@@ -1341,6 +1357,12 @@
 
 		if (arrival != 0)
 			(*(prot.prot->sleep_till))(&nextpkt_tv);
+#ifdef USEEVENTS
+		else
+			tgevent_poll();
+		if (gotevent)
+			break;
+#endif
 		lasttime = nextpkt_tv;
 
 		/* Did we exceed the limit on the number of packets to send? */
@@ -1382,6 +1404,10 @@
 						   pktlen,
 						   &(cur_tg->stop_before),
 						   &pktid);
+#ifdef USEEVENTS
+			if (gotevent)
+				break;
+#endif
 			}
 		}
 	}
@@ -2097,4 +2123,8 @@
 {
 
 	log_close();
+#ifdef USEEVENTS
+	tgevent_shutdown();
+#endif
+	exit(1);
 }
