diff -ur jove4.16.0.72.orig/insert.c jove4.16.0.72/insert.c
--- jove4.16.0.72.orig/insert.c	2002-03-20 14:49:38.000000000 -0700
+++ jove4.16.0.72/insert.c	2010-07-21 11:58:51.000000000 -0600
@@ -504,7 +504,7 @@
 		atchar = 0;
 	}
 
-	getline(atline->l_dline, genbuf);
+	jove_getline(atline->l_dline, genbuf);
 	atchar += tchar;
 	linecopy(genbuf, atchar, save);
 	atline->l_dline = putline(genbuf);
diff -ur jove4.16.0.72.orig/io.c jove4.16.0.72/io.c
--- jove4.16.0.72.orig/io.c	2010-04-15 23:41:46.000000000 -0600
+++ jove4.16.0.72/io.c	2010-07-21 11:58:51.000000000 -0600
@@ -1356,9 +1356,9 @@
 
 void
 #ifdef USE_PROTOTYPES
-getline proto((daddr addr, register char *buf))
+jove_getline proto((daddr addr, register char *buf))
 #else
-getline(addr, buf)
+jove_getline(addr, buf)
 daddr	addr;
 register char	*buf;
 #endif
diff -ur jove4.16.0.72.orig/io.h jove4.16.0.72/io.h
--- jove4.16.0.72.orig/io.h	2002-03-20 14:49:38.000000000 -0700
+++ jove4.16.0.72/io.h	2010-07-21 11:58:51.000000000 -0600
@@ -40,7 +40,7 @@
 	close_file proto((File *fp)),
 	d_cache_init proto((void)),
 	file_write proto((char *fname, bool app)),
-	getline proto((daddr addr, char *buf)),
+	jove_getline proto((daddr addr, char *buf)),
 	lsave proto((void)),
 	putreg proto((File *fp,LinePtr line1,int char1,LinePtr line2,int char2,bool makesure)),
 	read_file proto((char *file, bool is_insert)),
diff -ur jove4.16.0.72.orig/recover.c jove4.16.0.72/recover.c
--- jove4.16.0.72.orig/recover.c	2010-05-16 19:53:38.000000000 -0600
+++ jove4.16.0.72/recover.c	2010-07-21 11:58:51.000000000 -0600
@@ -166,7 +166,7 @@
 private char	*getblock proto((daddr atl));
 
 void
-getline(tl, buf)
+jove_getline(tl, buf)
 daddr	tl;
 char	*buf;
 {
@@ -597,7 +597,7 @@
 	Nchars = Nlines = 0L;
 	while (--nlines >= 0) {
 		addr = getaddr(ptrs_fp);
-		getline(addr, buf);
+		jove_getline(addr, buf);
 		Nlines += 1;
 		Nchars += 1 + strlen(buf);
 		fputs(buf, out);
diff -ur jove4.16.0.72.orig/temp.h jove4.16.0.72/temp.h
--- jove4.16.0.72.orig/temp.h	2002-03-20 14:49:39.000000000 -0700
+++ jove4.16.0.72/temp.h	2010-07-21 11:58:51.000000000 -0600
@@ -12,7 +12,7 @@
  * Yank and Kill are really easy; basically all we do is make copies
  * of the disk addresses of the lines (as opposed to the contents).
  * So, putline(char *buf) writes buf to the disk and returns a new
- * disk address.  getline(daddr addr, char *buf) is the opposite of
+ * disk address.  jove_getline(daddr addr, char *buf) is the opposite of
  * putline().  f_getputl(LinePtr line, File fp) reads from open fp
  * directly into the tmp file (into the buffer cache (see below))
  * and stores the address in line.  This is used during read_file to
diff -ur jove4.16.0.72.orig/util.c jove4.16.0.72/util.c
--- jove4.16.0.72.orig/util.c	2004-02-01 12:38:29.000000000 -0700
+++ jove4.16.0.72/util.c	2010-07-21 11:58:51.000000000 -0600
@@ -280,7 +280,7 @@
 			strcpy(buf, linebuf);
 		Jr_Len = strlen(linebuf);
 	} else
-		getline(line->l_dline, buf);
+		jove_getline(line->l_dline, buf);
 	return buf;
 }
 
diff -ur jove4.16.0.72.orig/util.h jove4.16.0.72/util.h
--- jove4.16.0.72.orig/util.h	2004-02-01 12:38:29.000000000 -0700
+++ jove4.16.0.72/util.h	2010-07-21 11:58:51.000000000 -0600
@@ -18,7 +18,7 @@
 #define eobp()		(lastp(curline) && eolp())
 #define eolp()		(linebuf[curchar] == '\0')
 #define firstp(line)	((line) == curbuf->b_first)
-#define getDOT()	getline(curline->l_dline, linebuf)
+#define getDOT()	jove_getline(curline->l_dline, linebuf)
 #define lastp(line)	((line) == curbuf->b_last)
 
 extern UnivPtr
