--- linux-2.6.23.i686/arch/x86/kernel/reboot.c~	2008-02-29 10:06:46.000000000 -0700
+++ linux-2.6.23.i686/arch/x86/kernel/reboot.c	2008-02-29 10:05:50.000000000 -0700
@@ -122,6 +122,15 @@
 			DMI_MATCH(DMI_BOARD_NAME, "0WF810"),
 		},
 	},
+        {       /* Handle problems with rebooting on Dell Optiplex 745's SFF*/
+                .callback = set_bios_reboot,
+		.ident = "Dell OptiPlex 745",
+		.matches = {
+                        DMI_MATCH(DMI_SYS_VENDOR, "Dell Inc."),
+			DMI_MATCH(DMI_PRODUCT_NAME, "OptiPlex 745"),
+			DMI_MATCH(DMI_BOARD_NAME, "0RF703"),
+                },
+        },
 	{	/* Handle problems with rebooting on Dell 2400's */
 		.callback = set_bios_reboot,
 		.ident = "Dell PowerEdge 2400",
