--- uClibc-0.9.30/extra/scripts/unifdef.c.orig	2010-02-25 11:13:25.000000000 -0700
+++ uClibc-0.9.30/extra/scripts/unifdef.c	2010-02-25 11:13:46.000000000 -0700
@@ -206,7 +206,7 @@
 static void             error(const char *);
 static int              findsym(const char *);
 static void             flushline(bool);
-static Linetype         getline(void);
+static Linetype         _getline(void);
 static Linetype         ifeval(const char **);
 static void             ignoreoff(void);
 static void             ignoreon(void);
@@ -512,7 +512,7 @@
 
 	for (;;) {
 		linenum++;
-		lineval = getline();
+		lineval = _getline();
 		trans_table[ifstate[depth]][lineval]();
 		debug("process %s -> %s depth %d",
 		    linetype_name[lineval],
@@ -526,7 +526,7 @@
  * help from skipcomment().
  */
 static Linetype
-getline(void)
+_getline(void)
 {
 	const char *cp;
 	int cursym;
